### ISSUE TITLE
Improve DynVar linking performance when reparenting

### ISSUE THREAD

**Issue by mpmxyz** [Yellow-Dog-Man/Resonite-Issues#3149]
### Is your feature request related to a problem? Please describe.

When an object with many dynamic variables is reparented each dynamic variable is updating its linking. This can cause noticable freezes when grabbing and dropping such objects.
An extreme case is the linking process of a large number of variables occupying the same slot. That is actually in `O(n²)` because component search in slots can iterate every component and it can do that for each individual linking process.

### Describe the solution you'd like

I cannot give sufficient advice on how to improve performance without the original source and proper profiling.
But I still have a few ideas based on how dynamic variables work currently:

The goal is to build a search tree made up of spaces and variables in sync with the world model. The tree only needs to be amended for the edges that are affected by reparenting. After that all potentially affected variables can be detected by recursion through the tree:
1. determine parent space
2. start at reparented slot
3. recursively go through hierarchy, use search tree instead of slots after discovering a space (-> potential to optimize away component search)
4. prune parts of the search tree where node is known to have no link via parent space (needs some kind of counter for each link going across a node)
  - This has the potential for major performance improvements. A simple object with a `DynamicVariableSpace` and a `Grabbable` within the same slot shields all contained variables not linking to a parent space from relinking.

A few implementation details I can imagine:
- Caching of (primary) dynamic variable space for each slot
  - When a slot is queried for a dynamic variable space the result is remembered for when it is requested again. (needs proper dirtying of cache)
  - This limits the number of times the bag of components of a slot need to be queried.
  - It could directly reduce the worst case of many variables occupying the same slot to be in `O(n)`.
- Chaining of dynamic variable spaces
  - Horizontal chaining (within same slot)
    - A primary dynamic variable space keeps track of all dynamic variable spaces in a slot.
    - This would allow per-slot caching with a single property even when multiple (named) dynamic variable spaces are within the same slot.
    - Alternatively the per-slot cache could be a list of dynamic variable spaces.
  - Vertical chaining (from child to parent)
    - Dynamic variables only link to the next available dynamic variable space which will be a proxy to its parent's and same-slot spaces.
    - This will replace skipping of differently named dynamic variable spaces by a link that is proxied to the correct space.
- Reparenting only triggers relinking for variables which are potentially affected
  1. Unlinked variables
  2. Linked variables which are linked across the gap between the .
  - This may need additional information in the direction from parent to child, like a set of child spaces and a set variables. (probably indexed by logical space)

### Describe alternatives you've considered

Creating alternatives to regular dynamic variables that trade features for performance have already been suggested. E.g. #3140 goes in the direction of giving access to unbound dynamic variables.
This could result in redundant systems or make existing features dependent on context.

A component that makes child hierarchies invisible to reparenting could also be a solution but introduces possible inconsistencies to the data model.

**Comment by Frooxius** [comment]
I appreciate you providing a lot of technical suggestions, but what we really need are more hard data on the problem first - we generally advise focusing on the high level, rather going to deep into technical stuff.

Notably the pieces that I find missing in this are some general data on the problem - 

- How many (roughly) dynamic variable spaces / dynamic variables are we talking? 
- And how long the freezes tend to be?
- How often are you actually re-parenting?
- Do you have replication steps / objects for this that we can test with and use for profiling?

**Comment by mpmxyz** [comment]
True, it affected me so directly that I forgot to take a step back from it.
My use-cases involves data collection and a tabular UI. The tabular UI has a main dynamic variable space per cell for content and several auxiliary spaces for tooltip information. (lets say 5 spaces per cell)
The table contains 13 rows that are moved around like a conveyor belt. (no change of parent)
Motion data of an animation can easily have 50 columns of information. That makes 5*50*13=3250 dynamic variable spaces with at least a hand full of dynamic variables each. (3250*5=16250)
A minute of motion data at 20 fps would contain 1200 frames with 60000 values. Since my tool is more generalized it is actually one dynamic variable space per value with a hand full of variables.
I agree that the motion data is excessive due to the lack of integrated data structures. So for the sake of profiling lets say we have an 2D array of 100x100 objects with a dynamic variable space and 7 variables each, one of which links to a parent space:
resrec:///U-TheAutopilot/R-4882c95e-535c-4008-92af-cea0c86b2510
https://github.com/user-attachments/assets/d83fb670-d7a6-46cf-8c5a-81734dc080df
This test demonstrates at least 45ms freeze when grabbing and at least 60ms freeze when releasing the test object with approx. 70k dynamic variables and 10k spaces.
